### PR TITLE
[Java] Bulk data transfer for Tensor class

### LINKED ISSUE #6576: [Java] Bulk data transfer for Tensor class
Enhance the `Tensor` class with bulk data-access methods, to support the following scenarios:
- serialization of tensors in managed code (regardless of tensor datatype)
- efficient transfer of large blocks of multidimensional data
- support for direct transfers (i.e. using direct `ByteBuffer`s)
- support for typed `Buffer`s

### PR DESCRIPTION
**Pull request by EronWright**
Closes #6576

Introduces bulk data-access pattern to `Tensor`, with tests.
- `Tensor::create(DataType dt, long[] shape, Buffer data)` method
- `Tensor::getDataByteSize()` method
- `Tensor::readData(Buffer data)` method

Details:
- rename `allocate` to `allocateNDArray` since it is designed for the
NDarray scenario.
- introduce a native `allocate` method that accepts the data size in
bytes, rather than computing the size with shape information (which
isn’t always sufficient, e.g. TF_STRING).
- introduce a private `buffer` accessor to access the TF data as a
direct `ByteBuffer`.
- implement `create` and `readData` methods to read and write data
using a variety of `Buffer` variants.   Supports typed buffers and
direct buffers.

### CODE REVIEW

**Comment by EronWright** [comment]
@asimshankar please have a look, thanks.

**Comment by asimshankar** [comment]
Thanks for the PR! Took a quick look, and at a very high level this looks great. Will make a detailed pass a little later (apologies for the slow responses, I've been traveling but will be back next week and faster to respond thereafter :)

One thing to keep in mind: Since part of the intention here is for serialization, this isn't quite accurate. Turns out that the in-memory representation and serialized form aren't always the same (even if you ignore endian-ness).  Ideally we'd always use what is filled in [`TensorProto`](https://github.com/tensorflow/tensorflow/blob/9c41d5f64efedf44d30e5cc21f900ea07097395c/tensorflow/core/framework/tensor.proto#L36) as the serialized form, which is produced by [`Tensor::AsProtoTensorContent`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/tensor#asprototensorcontent) in C++. 

So I would avoid committing to the `ByteBuffer` over the wire. Do keep that in mind. Though, at this stage I'm mostly interested in getting the API right, not having a comprehensive implementation in the short term is fine (For example, we did that with [Go](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/go/tensor.go#L196)).

**Comment by EronWright** [comment]
I agree that TensorProto would be ideal for serialization, and spent some time on it.  But became discouraged when I realized that the generated classes don't use primitive arrays. For many scenarios, Buffers provide the ideal performance and flexibility.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: I'm a bit weary about all these `instanceof` statements and the possibility of `throwIncompatibleBuffer`. How about we make this just:

```java
public void copyTo(ByteBuffer dst);
```

Callers can use `dst.asFloatBuffer()` etc. as desired. This will also be consistent with the fact that we only provide guidance on the argument to `ByteBuffer.allocate()` (`getDataByteSize()`) and no direct guidance on the argument to `FloatBuffer.allocate()` etc.

File: tensorflow/java/src/main/native/tensor_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/tensor_jni.cc]: Most of this is duplicated in `allocateNDArray`, perhaps structure this so that the code isn't duplicated (`allocateNDArray` can call `allocate` after determining the byte size). Though, see comment above about `allocate` and `allocateForStrings`.

File: tensorflow/java/src/main/native/tensor_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/tensor_jni.cc]: Nits:
- Use `static_cast<jlong>` instead of c-style casts
- Perhaps just `return env->NewDirectByteBuffer(...)`

File: tensorflow/java/src/test/java/org/tensorflow/TensorTest.java
Comment (asimshankar) [inline tensorflow/java/src/test/java/org/tensorflow/TensorTest.java]: Not all of these are re-used between tests (for example, the shape ones are needed only for the `create` tests).

I'd suggest doing away with class variables and instead defining these within the test methods. That may come at the cost of some duplication, but the resulting tests are more readable if they are self contained.

File: tensorflow/java/src/test/java/org/tensorflow/TensorTest.java
Comment (asimshankar) [inline tensorflow/java/src/test/java/org/tensorflow/TensorTest.java]: Let's use the [`try-with-resources`](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) statement here (which can handle multiple resources within the `try`):

```java
try (Tensor tscalar = Tensor.create(scalar);
       Tensor tvector = Tensor.create(vector);
       ....) {
   // Body of the test
}
```

That way you don't need the `finally` clause, don't have to worry about the `finally` clause covering all the tensors and it also takes care of any errors within the `create` calls.

File: tensorflow/java/src/test/java/org/tensorflow/TestUtil.java
Comment (asimshankar) [inline tensorflow/java/src/test/java/org/tensorflow/TestUtil.java]: I feel the two `flatten` methods warrant some comments :)

**Review by asimshankar — COMMENTED** [review]
Thanks for the PR! Generally looks good, but I have some API comments.

Apologies for the delay, I was traveling, but am back now.

Thanks!

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Or maybe just have a single:

```java
private static native long allocate(int dtype, long[] shape, long sizeInBytes);
```

Since it is a private method, we can make it require that `sizeInBytes` be consistent with `shape` and `dtype`. So we don't need two methods, just one?

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Believe it or not, my first draft was entirely `ByteBuffer` based.   I found it necessary to support other `Buffer` types to optimize the transfer of wrapped primitive arrays. 

Consider a `FloatBuffer` as produced by `FloatBuffer.wrap(new float[] { ... })`.  It is not convertible to a `ByteBuffer`.   To write it directly to the tensor's native `ByteBuffer`, we call `put` on a view of the latter.   At all times we avoid exposing the native buffer to user code.

I'm fine with moving to a set of overloaded methods to retain the benefits of other buffer types while reducing the possibility of runtime exceptions.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: I imagined the term `read` to have `InputStream` semantics, but I hear you.

My preference would be `writeTo(Buffer)`, but can do `copyTo` if you prefer.   I'm personally not thrilled by how open-ended `copyTo` is, that's why I like the former.

Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Keep in mind that `FloatBuffer`s that wrap an array are writable, so `copyTo` applied to a float buffer is more efficient than it would be if applied to a `ByteBuffer` (assuming you ultimately want a `float[]`).

I agree that the lack of guidance on sizing of typed buffers is inconvenient, but some utility methods might be enough.

Keeping with the power-over-convenience assumption, I'm inclined to provide optimized transfers.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: I agree with your latter point.  If I recall correctly, it would require that the element counting code within the ND array codepath be moved from C to Java.  Since we have `flattenedNumElements` in the managed test code anyway, I have no objection.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Okay, let's do the overloaded methods then.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Let's do `writeTo(Buffer)`

Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Alright, I guess we keep the `instanceof` checks for now.
One minor thing, instead of `throwIncompatibleBuffer()`, mind make the helper function:

```java
private RuntimeException incompatibleBuffer(String got, String want) {
  throw new IllegalArgumentException(...)
}
```

And then at call sites do:
`throw incompatibleBuffer("Byte", "Float")`
(or something along those lines),

so that the stack trace is shorter and some more detail is provided.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Doing the flattened counting in Java should be fine. So let's do that.

**Comment by drpngx** [comment]
Any luck with this @EronWright ?

**Comment by EronWright** [comment]
@drpngx ETA tomorrow.

**Comment by drpngx** [comment]
No hurry, just checking. Nice contribution!

**Comment by drpngx** [comment]
@asimshankar Just to be sure - is that good to go?

**Comment by EronWright** [comment]
A comment on why the typed `create` methods take a `DataType` parameter, it is to support unsigned values in the future.   i.e.:
```java
IntBuffer buffer = ...
Tensor t = Tensor.create(DataType.UINT32, shape, buffer);
```

Note the JDK 8 support for unsigned values ([reference](https://blogs.oracle.com/darcy/entry/unsigned_api)).

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Nit: Can you place all these specialized `create()` methods after the one that creates from an `Object` so that all the `create()` methods are grouped together.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Nit: Perhaps call this `createHelper` or something else just to reinforce the fact that this is slightly different from all the other `create()` methods (it doesn't initialize the contents)

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Nit: Remove this line.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Perhaps can make this even more informative with:

```java
private static IllegalArgumentException incompatibleBuffer(Buffer buf, DataType dataType) {
  return new IllegalArgumentException("cannot use " + buf.getClass().getName() + " with Tensor of type " + dataType);
}
```

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Same here:

```java
private static IllegalArgumentException incompatibleBuffer(int numElements, long[] shape) {
  return new IllegalArgumentException("buffer with " + numElements + " elements is not compatible with a Tensor that has a shape " + Arrays.toString(shape));
}
```

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Remove the quotes and expand on the message: 
```java
throw new IllegalArgumentException("unrecognized DataType " + dataType);
```

File: tensorflow/java/src/main/native/tensor_jni.cc
Comment (asimshankar) [inline tensorflow/java/src/main/native/tensor_jni.cc]:
```c++
static_cast<size_t>(sizeInBytes)
```

File: tensorflow/java/src/main/native/tensor_jni.h
Comment (asimshankar) [inline tensorflow/java/src/main/native/tensor_jni.h]: Nit: Place this method in the `.cc` and `.h` files after `allocate` and `delete`

File: tensorflow/java/src/test/java/org/tensorflow/TensorTest.java
Comment (asimshankar) [inline tensorflow/java/src/test/java/org/tensorflow/TensorTest.java]: Why is the `@SuppressWarnings` decorator needed?

File: tensorflow/java/src/test/java/org/tensorflow/TensorTest.java
Comment (asimshankar) [inline tensorflow/java/src/test/java/org/tensorflow/TensorTest.java]: Would it make sense to split these into separate tests?

1. "byte" tests - like use of the direct buffer and the endian-ness flips

2. One for all the typed buffers (lines 68-91)

3. One for shape checking (lines 92-128)

4. One for datatype checking

File: tensorflow/java/src/test/java/org/tensorflow/TensorTest.java
Comment (asimshankar) [inline tensorflow/java/src/test/java/org/tensorflow/TensorTest.java]: If possible, I'd suggest breaking this test too into one per type.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Perhaps call this `numBytes()`, so we will have `numBytes()` and `numElements()`

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: In all the `writeTo` methods, maybe check consistency before calling `buffer()`?

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: It may just be clearer to return constants here. The byte sizes for TensorFlow data types are fixed, so let's return those?

```java
case FLOAT:
case INT32:
   return 4;
case DOUBLE:
case INT64:
  return 8;
case BOOL:
  return 1;
```

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (asimshankar) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Place this after `allocate` and `delete`

**Review by asimshankar — COMMENTED** [review]
Thanks for the update, I still suggest that we do without the `DataType` argument. And then some other minor things.

**Comment by EronWright** [comment]
I plan to work on these changes this weekend.

**Comment by asimshankar** [comment]
@EronWright : Any updates? We're looking forward to this :)

**Comment by EronWright** [comment]
@asimshankar sorry for the delay, I need a few more days.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: Was unable to do this, tried to but the `ByteBuffer` variant was a special case.

File: tensorflow/java/src/main/java/org/tensorflow/Tensor.java
Comment (EronWright) [inline tensorflow/java/src/main/java/org/tensorflow/Tensor.java]: I do want to learn, and fixed the catch blocks accordingly.

File: tensorflow/java/src/test/java/org/tensorflow/TensorTest.java
Comment (EronWright) [inline tensorflow/java/src/test/java/org/tensorflow/TensorTest.java]: I reworked it somewhat, separating out the `ByteBuffer` tests.

**Comment by EronWright** [comment]
@asimshankar ready for you to have another look.  Thanks again.

**Review by asimshankar — APPROVED** [review]
Thanks very much for the contribution @EronWright ! This is sincerely appreciated.

Jenkins, test this please.

This should be ready to merge once tests pass.  I think the API is what we want it to be.

There are some minor style tweaks (a space here, a comma there), but not worth holding up this pull request. So let's merge it in and I'll make those cosmetic tweaks later on.

Thanks again!

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
This fails in `//tensorflow/python:coordinator_test`. @rmlarsen, feel free to merge if this is a known failure.

**Comment by vrv** [comment]
@asimshankar I believe GitHub now lets you edit PRs, so you can go ahead and make small typo changes yourself :)

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
It appears that the test failures are unrelated. Merging.

**Pull request merged**
